services/worker-manager/src/providers/google.js

```diff
@@ -354,58 +354,51 @@ class GoogleProvider extends Provider {
     this.seen[worker.workerPoolId] = this.seen[worker.workerPoolId] || 0;
     this.errors[worker.workerPoolId] = this.errors[worker.workerPoolId] || [];
 
-    let deleteOp = false;
-    if (worker.providerData.operation) {
-      deleteOp = await this.handleOperation({
-        op: worker.providerData.operation,
-        errors: this.errors[worker.workerPoolId],
-      });
-    }
-
-    let res;
     try {
-      res = await this._enqueue('get', () => this.compute.instances.get({
+      const {data} = await this._enqueue('get', () => this.compute.instances.get({
         project: worker.providerData.project,
         zone: worker.providerData.zone,
         instance: worker.workerId,
       }));
+      const {status} = data;
+      if (['PROVISIONING', 'STAGING', 'RUNNING'].includes(status)) {
+        this.seen[worker.workerPoolId] += 1;
+
+        // If the worker will be expired soon but it still exists,
+        // update it to stick around a while longer. If this doesn't happen,
+        // long-lived instances become orphaned from the provider. We don't update
+        // this on every loop just to avoid the extra work when not needed
+        if (worker.expires < taskcluster.fromNow('1 day')) {
+          await worker.modify(w => {
+            w.expires = taskcluster.fromNow('1 week');
+          });
+        }
+      } else if (['TERMINATED', 'STOPPED'].includes(status)) {
+        await this._enqueue('query', () => this.compute.instances.delete({
+          project: worker.providerData.project,
+          zone: worker.providerData.zone,
+          instance: worker.workerId,
+        }));
+        await worker.modify(w => {
+          w.state = states.STOPPED;
+        });
+      }
     } catch (err) {
       if (err.code !== 404) {
         throw err;
       }
-      await worker.modify(w => {
-        w.state = states.STOPPED;
-        if (deleteOp) {
-          delete w.providerData.operation;
-        }
-      });
-      return;
-    }
-    const {status} = res.data;
-    if (['PROVISIONING', 'STAGING', 'RUNNING'].includes(status)) {
-      this.seen[worker.workerPoolId] += 1;
-
-      // If the worker will be expired soon but it still exists,
-      // update it to stick around a while longer. If this doesn't happen,
-      // long-lived instances become orphaned from the provider. We don't update
-      // this on every loop just to avoid the extra work when not needed
-      if (worker.expires < taskcluster.fromNow('1 day')) {
-        await worker.modify(w => {
-          w.expires = taskcluster.fromNow('1 week');
+      if (worker.providerData.operation) {
+        // We only check in on the operation if the worker failed to
+        // start succesfully
```

Comment:
..and this knows that the start failed because the `compute.instance.get` returned a 404?

Reply:
That's my working model, yes. If that proves to not be true later then we can go back to the complicated way. I'm thinking that if the instance was created then the operation that reports the status of creating the instance will not have useful info on a failure.